lib/DxcSupport/HLSLOptions.cpp

```diff
@@ -588,7 +588,35 @@ int ReadDxcOpts(const OptTable *optionTable, unsigned flagsToInclude,
   opts.UseInstructionNumbers = Args.hasFlag(OPT_Ni, OPT_INVALID, false);
   opts.UseInstructionByteOffsets = Args.hasFlag(OPT_No, OPT_INVALID, false);
   opts.UseHexLiterals = Args.hasFlag(OPT_Lx, OPT_INVALID, false);
-  opts.Preprocess = Args.getLastArgValue(OPT_P);
+  if (Args.hasFlag(OPT_P, OPT_INVALID, false)) {
+    // Default preprocess filename is InputName.i.
+    llvm::SmallString<128> Path(Args.getLastArgValue(OPT_INPUT));
+    llvm::sys::path::replace_extension(Path, "i");
+    // Try to get preprocess filename from Fi.
+    opts.Preprocess = Args.getLastArgValue(OPT_Fi, Path).str();
+    // Hack to support fxc style /P preprocess_filename.
+    // When there're more than 1 Input file, use the input which is after /P as
+    // preprocess.
+    std::vector<std::string> Inputs = Args.getAllArgValues(OPT_INPUT);
+    if (Inputs.size() > 1) {
+      llvm::opt::Arg *PArg = Args.getLastArg(OPT_P);
+      std::string LastInput = Inputs.back();
+      llvm::opt::Arg *PrevInputArg = nullptr;
+      for (llvm::opt::Arg *InputArg : Args.filtered(OPT_INPUT)) {
+        // Find Input after /P.
+        if ((PArg->getIndex() + 1) == InputArg->getIndex()) {
+          opts.Preprocess = InputArg->getValue();
+          if (LastInput == opts.Preprocess && PrevInputArg) {
+            // When InputArg is last Input, update it to other Input so
+            // Args.getLastArgValue(OPT_INPUT) get expect Input.
+            InputArg->getValues()[0] = PrevInputArg->getValues()[0];
+          }
+          break;
+        }
+        PrevInputArg = InputArg;
```

Comment:
Overall, this is kind of ugly, but I don't have a better alternative and I like that it maintains the previous behavior. We might consider including a warning suggesting using the new interface. I realize that probably contradicts something I said before. Sorry :(